### PR TITLE
#BN-7 feat: Create Roles and Set Permissions

### PR DESCRIPTION
**Pull request by kunda4**
# PR Description
Admin should be able to assign various users to various roles. There should also be a role for the admin account itself. This will allow only accounts with particular roles permission to peform particular actions.
# Description of tasks that were expected to be completed
List all the tasks and sub-tasks that were assigned to you Make sure you check every completed task.
- [x] admin can create roles
- [x] admin assigns a particular user to a particular role
- [x] the user can only perform actions allowed by their role.
- [x] Document feature using swagger
OPTIONAL FEATURE (CHECK WITH MANAGER)
- [x] Each role should have a list of permissions attached to them.
- [x] Admin should be able to add and remove permissions from particular roles.
# How this may be tested?
Please describe the tests that you ran to verify your functionalities. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
1. run `npm i` to install all required dependencies
2. run `npm run dev`
3. step  use postman to test endpoint or swagger
4. step 4...
# Screenshots (If appropriate/applicable)
Provide screenshots related to your PR in case it's appropriate and applicable
# Please check this Checklist before you submit your PR:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My code generates no warnings
- [x] There are no vulnerabilities

### CODE REVIEW

File: src/controllers/permController.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        const { permName, description } = req.body;
+        try {
+            if (!Permission) {
+                return res.status(404).json({
+                    statusCode: 404,
+                    Message: req.t('id_unexist_message'),
+                });
+            } else {              
+                await permission.update(
+                    { permName, description },
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Better to have permName as permssionName

File: src/controllers/permController.js
Comment (chrisostome01) [inline src/controllers/permController.js:1]: Please consider formatting this file eslint prettier

File: src/controllers/permController.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
+
+import {permission,rolepermission} from "../database/models";
+    
+    const createNewPermission = async(req, res)=> {
+    const { permName, description } = req.body; 
+        try {
+            const newPermission = await permission.create({permName, description})
+            return res
+        .status(201)
+        .json({ message: req.t('permission_create_message'), data: newPermission });
```
Comment (chrisostome01) [inline src/controllers/permController.js]: remember to include you statusCode in the response too

File: src/controllers/permController.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        } 
+        catch (err) {
+            console.log('error ' + err);
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Remove this console

File: src/controllers/permController.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
+
+import {permission,rolepermission} from "../database/models";
+    
+    const createNewPermission = async(req, res)=> {
+    const { permName, description } = req.body;
```
Comment (chrisostome01) [inline src/controllers/permController.js]: PermName, should be meaningful.

File: src/controllers/permController.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        catch (err) {
+            console.log('error ' + err);
+            return res.status(500).json(err);
+        }     
+    }
+
+    const getAllPermission = async(req, res)=> {
+        try {
+            const permissions = await permission.findAll();
+            res.status(200).json({ statusCode: 200, data: permissions });
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Message is missing in your response

File: src/controllers/permController.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    const getOnePermission = async(req, res)=> {
+        const id = req.params.id;
+        console.log(id);
+        try {
+            const permissions = await permission.findOne({
+                where: { id },
+            });
+            if (permissions) {
+                return res.status(200).json({ statusCode: 200, data: permissions });
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Message is missing here

File: src/controllers/permController.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                });
+                res.status(200).json({
+                    statusCode: 200,
+                    message: req.t('permission_delete_message'),
+                    
+                });
+            } else {
+                res.status(404).json({
+                    statusCode: 404,
+                    Message: req.t('id_unexist_message'),
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Stay consistent `Message` is completely different from `message`

File: src/controllers/roleController.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        const roleData = {
+            userId: req.body.userId,
+            roleId: req.body.roleId,
+        }
+        if (!userId) {
+            return res.status(404).send(`${userId} is not found`);
+          }
+        try {
+            const userRoles = await userRole.create(roleData)
+            return res.status(201).json({ message: req.t('set-role-message'), response: userRoles })
```
Comment (chrisostome01) [inline src/controllers/roleController.js]: snake case would be good instead `set-role-message` you can use `set_role_message`

File: src/controllers/roleController.js [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+                await userRole.update(
+                    { userId, roleId },
+                    {
+                        where: {id},});
+                return res
+                    .status(201)
+                    .json({ statusCode: 201, message: req.t('update_message'),data:UserWithRole });
+            }
+        } catch (error) {
+            console.error(error);
```
Comment (chrisostome01) [inline src/controllers/roleController.js]: remove this console

File: src/controllers/roleController.js [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+            return res
+                .status(500)
+                .json({ statusCode: 400, error: "Server error" });
```
Comment (chrisostome01) [inline src/controllers/roleController.js]: status code are deferent...

File: src/controllers/userController.js [hunk 11]
```diff
@@ -73,9 +73,9 @@ export async function loginUser(req, res) {
 
 export async function createUser(req, res) {
   const {
-    firstName, lastName, email, password, phone, role,
+    firstName, lastName, email, password, phone, roleId,
   } = req.body;
-  if (!firstName || !lastName || !email || !password || !phone || !role) {
+  if (!firstName || !lastName || !email || !password || !phone || !roleId) {
```
Comment (chrisostome01) [inline src/controllers/userController.js]: Use Joi to validate all of those inputs

File: src/controllers/permController.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+                res.status(200).json({ statusCode: 200, data: RoleWithPermission });
+            }
+             else {
+                res.status(404).json({
+                    statusCode: 404,
+                    Message: req.t('id_unexist_message'),
+                });
+            }
+        } catch (error) {
+            return res.status(500).json({ error: error.message })
```
Comment (chrisostome01) [inline src/controllers/permController.js]: Use message here instead of error to improve consistency

**Review by chrisostome01 — CHANGES_REQUESTED** [review]
# Well done @kunda4 the codes are really well structured however here are a couple of things that we need to work on.

- Improve consistency in your code structure: some lines of code are written differently but in a different way but the same context. `message` and `Message` or `error` instead of `message`
- You can generate your own seeds instead of reusing the same file.
- Please make sure you are validating all of the data you are receiving 
- Please format the code for more structured codes

File: Makefile [hunk 13]
```diff
@@ -19,4 +19,8 @@ migrate-rollback:
 seed:
 	docker exec e_commerce-node-app-1 npm run seed
 seed-rollback:
-	docker exec e_commerce-node-app-1 nom run seed-down
\ No newline at end of file
+<<<<<<< HEAD
+	docker exec e_commerce-node-app-1 nom run seed-down
+=======
+	docker exec e_commerce-node-app-1 nom run seed-down
+>>>>>>> a84c243cf595935bb3c658307b81e93305ae41fb
```
Comment (chrisostome01) [inline Makefile]: I those conflicts are not supposed to be here..?

File: package.json [hunk 14]
```diff
@@ -4,9 +4,11 @@
   "description": "",
   "main": "index.js",
   "scripts": {
+    "make-model": "npx sequelize-cli model:generate --name users --attributes firstName:string,lastName:string,email:string, password:string, role:string",
```
Comment (chrisostome01) [inline package.json]: Remove those demo data `users firstName:string,lastName:string,email:string, password:string, role:string`

File: package.json [hunk 15]
```diff
@@ -4,9 +4,11 @@
...
     "migrate": "npx sequelize-cli db:migrate",
     "seed": "npx sequelize-cli db:seed:all",
     "seed-down": "npx sequelize-cli db:seed:undo:all",
+    "make-seed": "npx sequelize-cli seed:generate --name demo-users",
```
Comment (chrisostome01) [inline package.json]: remove also this `--name demo-users` this should be someone else's inputs

File: src/controllers/permController.js
Comment (chrisostome01) [reply src/controllers/permController.js:1]: permssionController would be a good choice for this fileName and it would be more descriptive enough

**Review by chrisostome01 — CHANGES_REQUESTED** [review]

**Review by Honore123 — APPROVED** [review]
🚀

**Review by chrisostome01 — APPROVED** [review]
# LGTM

**Pull request merged**